modules/internal/ChapelArray.chpl

```diff
@@ -2625,10 +2629,20 @@ module ChapelArray {
     /* The number of dimensions in the array */
     proc rank param return this.domain.rank;
 
-    /* return the array's indices as its domain */
-    pragma "return not owned"
-    proc indices
+    proc indices where arrayIndicesAlwaysLocal == false {
+      compilerWarning("In a forthcoming release, calling '.indices' on an array will return/yield its indices locally rather than returning a copy of the array's domain as it does today.  See https://github.com/chapel-lang/chapel/issues/17883 for details.  To preserve the current behavior, use '.domain' or a copy thereof.  To update to the proposed new behavior today, re-compile with '-sarrayIndicesAlwaysLocal=true'.");
       return _dom;
+    }
+    
+    /* return the array's indices as its domain */
+    proc indices where arrayIndicesAlwaysLocal == true && !isSparseArr(this) && !isAssociativeArr(this) {
+      return {(..._dom.getIndices())};
+    }
+
+    iter indices where arrayIndicesAlwaysLocal == true && (isSparseArr(this) || isAssociativeArr(this)) {
```

Comment:
A practical challenge related to the 3rd bullet in https://github.com/chapel-lang/chapel/issues/18353: how do we document this?

But for now, should we `no doc` this, and add a note to the one above? Or add a documentation for this, and explain for both this and the one above which one's supported where?